src/map_memory.h

```diff
@@ -0,0 +1,44 @@
+#pragma once
+#ifndef MAP_MEMORY_H
+#define MAP_MEMORY_H
+
+#include "enums.h"
+
+#include <list>
+#include <string>
+#include <unordered_map>
+
+class JsonOut;
+class JsonObject;
+
+struct memorized_terrain_tile {
+    std::string tile;
+    int subtile;
+    int rotation;
+};
+
+class map_memory
+{
+    public:
+        void store( JsonOut &jsout ) const;
+        void load( JsonObject &jsin );
+
+        /** Memorizes a given tile; finalize_tile_memory needs to be called after it */
+        void memorize_tile( int limit, const tripoint &pos, const std::string &ter,
+                            const int subtile, const int rotation );
+        /** Returns last stored map tile in given location */
+        memorized_terrain_tile get_tile( const tripoint &p ) const;
+
+        void memorize_symbol( int limit, const tripoint &pos, const long symbol );
+        long get_symbol( const tripoint &p ) const;
+    private:
+        void trim( int limit );
```

Comment:
Since the value of limit is unlikely to change repeatedly during any given game session, is there any advantage to reserving the lists and unordered_map to limit size, caching the value of limit in map_memory, and then resizing the lists and unordered_maps if limit changes?

Reply:
> You can't reserve anything in a `std::list` unless a custom allocator is used. Even if it would be possible, this won't help here since `std::list` doesn't allow manipulating its nodes directly (like taking a node in the middle and putting it in front without heap allocations).

You can move nodes around without heap allocations using `splice`, but indeed there's no way to extract a node and have it as a separate object.

Reply:
I expect this container to fill up once and then just churn, so I dont think the savings from trying to presize it is significant.
I abandoned storing limit in this class because it made initialization weird.